service/rtc/sfu.go

```diff
@@ -354,42 +408,93 @@ func (s *Server) InitSession(cfg SessionConfig, closeCb func() error) error {
 
 			s.log.Debug("received screen sharing stream", mlog.String("streamID", streamID), mlog.String("sessionID", us.cfg.SessionID))
 
-			outScreenTrack, err := webrtc.NewTrackLocalStaticRTP(rtpVideoCodecVP8, genTrackID("screen", us.cfg.SessionID), random.NewID())
+			outScreenTrack, err := webrtc.NewTrackLocalStaticRTP(params.RTPCodecCapability, genTrackID("screen", us.cfg.SessionID), random.NewID(), webrtc.WithRTPStreamID(remoteTrack.RID()))
 			if err != nil {
 				s.log.Error("failed to create local track",
 					mlog.Err(err), mlog.String("sessionID", us.cfg.SessionID))
 				return
 			}
+
+			rid := remoteTrack.RID()
+			if rid == "" {
+				rid = SimulcastLevelDefault
+			}
+
+			rm, err := NewRateMonitor(2*time.Second, nil)
+			if err != nil {
+				s.log.Error("failed to create rate monitor", mlog.Err(err))
+				return
+			}
+
 			us.mut.Lock()
-			us.outScreenTrack = outScreenTrack
-			us.remoteScreenTrack = remoteTrack
+			us.outScreenTracks[rid] = outScreenTrack
+			us.remoteScreenTracks[rid] = remoteTrack
+			us.screenRateMonitors[rid] = rm
 			us.mut.Unlock()
 
 			call.iterSessions(func(ss *session) {
 				if ss.cfg.UserID == us.cfg.UserID {
 					return
 				}
+
+				expectedLevel := SimulcastLevelDefault
+				if remoteTrack.RID() != "" {
+					expectedLevel = ss.getExpectedSimulcastLevel()
+				}
+
+				if rid != expectedLevel {
+					return
+				}
+
+				s.log.Debug("received track matches expected level, sending",
+					mlog.String("lvl", expectedLevel),
+					mlog.String("sessionID", us.cfg.SessionID),
+				)
+
 				select {
-				case ss.tracksCh <- outScreenTrack:
+				case ss.tracksCh <- trackActionContext{action: trackActionAdd, track: outScreenTrack}:
 				default:
 					s.log.Error("failed to send screen track: channel is full",
-						mlog.String("UserID", us.cfg.UserID),
+						mlog.String("userID", us.cfg.UserID),
 						mlog.String("sessionID", us.cfg.SessionID),
 						mlog.String("trackUserID", ss.cfg.UserID),
 						mlog.String("trackSessionID", ss.cfg.SessionID),
 					)
 				}
 			})
 
+			limiter := rate.NewLimiter(0.25, 1)
+
 			for {
-				rtp, _, readErr := remoteTrack.ReadRTP()
+				buf := s.bufPool.Get().([]byte)
+				i, _, readErr := remoteTrack.Read(buf)
```

Comment:
iiuc, this is to get the number of bytes? Or is there another reason? If it's the former, then isn't something like `len(rtp.payload)` easier/faster?

Reply:
Yes, we need the number of received bytes to feed it to the monitor for calculating the source rate.

> If it's the former, then isn't something like len(rtp.payload) easier/faster?

That would only return the size of the payload, excluding any header and padding. To be accurate we want to pass everything we get since it's going to be relayed to the peers eventually.